internal/backend/db/dbgorm/projects.go

```diff
@@ -41,24 +81,61 @@ func (db *gormdb) UpdateProject(ctx context.Context, p sdktypes.Project) error {
 	return nil
 }
 
+func (db *gormdb) getProject(ctx context.Context, projectID string) (*scheme.Project, error) {
+	return getOne(db.db, ctx, scheme.Project{}, "project_id = ?", projectID)
+}
+
+func (db *gormdb) getProjectByName(ctx context.Context, projectName string) (*scheme.Project, error) {
+	return getOne(db.db, ctx, scheme.Project{}, "name = ?", projectName)
+}
+
+func schemaToSDKProject(p *scheme.Project, err error) (sdktypes.Project, error) {
+	if p == nil || err != nil {
+		return sdktypes.InvalidProject, translateError(err)
+	}
+	return scheme.ParseProject(*p)
+}
+
 func (db *gormdb) GetProjectByID(ctx context.Context, pid sdktypes.ProjectID) (sdktypes.Project, error) {
-	return getOneWTransform(db.db, ctx, scheme.ParseProject, "project_id = ?", pid.String())
+	return schemaToSDKProject(db.getProject(ctx, pid.String()))
 }
 
 func (db *gormdb) GetProjectByName(ctx context.Context, ph sdktypes.Symbol) (sdktypes.Project, error) {
-	return getOneWTransform(db.db, ctx, scheme.ParseProject, "name = ?", ph.String())
+	return schemaToSDKProject(db.getProjectByName(ctx, ph.String()))
 }
 
-func (db *gormdb) ListProjects(ctx context.Context) ([]sdktypes.Project, error) {
-	var rs []scheme.Project
-	q := db.db.WithContext(ctx)
+type DepEnv struct {
+	DeploymentID string `gorm:"primaryKey"`
+	State        int32
+	EnvID        string
+}
 
-	err := q.Order("project_id").Find(&rs).Error
-	if err != nil {
-		return nil, translateError(err)
+func (db *gormdb) getProjectDeployments(ctx context.Context, pid string) ([]DepEnv, error) {
+	var pds []DepEnv
+	res := db.db.WithContext(ctx).Model(&scheme.Deployment{}).
+		Joins("join Envs on Envs.env_id = Deployments.env_id").
+		Where("Envs.project_id = ?", pid).
+		Select("Deployments.deployment_id, Deployments.state, Envs.env_id").
+		Find(&pds)
+	return pds, res.Error
+}
+
+func (db *gormdb) listProjects(ctx context.Context) ([]scheme.Project, error) {
+	q := db.db.WithContext(ctx).Order("project_id")
```

Comment:
do you really need to sort by project_id here ?
anyhow, I think it's the default sort anyway so it has no meaning

Reply:
I kept the existing code logic. Not added by me, so prefer not to remove it